crates/sui-core/src/transaction_signing_filter.rs

```diff
@@ -0,0 +1,171 @@
+            "Access to account address {:?} is temporarily disabled",
+            tx_data.sender()
+        )
+    );
+    deny_if_true!(
+        deny_map.contains(&tx_data.gas_owner()),
+        format!(
+            "Access to account address {:?} is temporarily disabled",
+            tx_data.gas_owner()
+        )
+    );
+    Ok(())
+}
+
+fn check_input_objects(
+    filter_config: &TransactionDenyConfig,
+    input_objects: &[InputObjectKind],
+) -> SuiResult {
+    let deny_map = filter_config.get_object_deny_map();
+    let shared_object_disabled = filter_config.shared_object_disabled();
+    if deny_map.is_empty() && !shared_object_disabled {
+        // No need to iterate through the input objects if no relevant policy is set.
+        return Ok(());
+    }
+    for object_kind in input_objects {
+        let id = object_kind.object_id();
+        deny_if_true!(
+            deny_map.contains(&id),
+            format!("Access to input object {:?} is temporarily disabled", id)
+        );
+        deny_if_true!(
+            shared_object_disabled && object_kind.is_shared_object(),
+            "Usage of shared object in transactions is temporarily disabled"
+        );
+    }
+    Ok(())
+}
+
+fn check_package_dependencies(
+    filter_config: &TransactionDenyConfig,
+    tx_data: &TransactionData,
+    package_store: &impl BackingPackageStore,
+) -> SuiResult {
+    let deny_map = filter_config.get_package_deny_map();
+    if deny_map.is_empty() {
+        return Ok(());
+    }
+    let mut dependencies = vec![];
+    for command in tx_data.kind().iter_commands() {
+        match command {
+            Command::Publish(_, deps) => {
+                // It is possible that the deps list is inaccurate since it's provided
+                // by the user. But that's OK because this publish transaction will fail
+                // to execute in the end. Similar reasoning for Upgrade.
+                dependencies.extend(deps.iter().copied());
+            }
+            Command::Upgrade(_, deps, package_id, _) => {
+                dependencies.extend(deps.iter().copied());
+                // It's crucial that we don't allow upgrading a package in the deny list,
+                // otherwise one can bypass the deny list by upgrading a package.
+                dependencies.push(*package_id);
+            }
+            Command::MoveCall(call) => {
+                let package =
+                    package_store
+                        .get_package(&call.package)?
+                        .ok_or(SuiError::UserInputError {
+                            error: UserInputError::ObjectNotFound {
+                                object_id: call.package,
+                                version: None,
+                            },
+                        })?;
+                // linkage_table contains all transitive dependencies of the package.
+                dependencies.extend(package.linkage_table().keys().copied());
```

Comment:
You don't want the `keys` here, you want the `upgrade_id` in the values -- this is the ID of the package in storage that actually gets loaded during execution (and corresponds to the value that comes from `deps` in the `Publish` and `Upgrade` transactions).

Having said that, maybe you want to be able to prevent calls to all versions of a package?  In which case you should use the `keys()` of the linkage table and the `original_package_id()` of the package being called as well.